test/acvp_kyber.c

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: Apache-2.0
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include "kem.h"
+#include "randombytes.h"
+
+#define USAGE \
+  "acvp_kyber{lvl} [encapDecap|keyGen] [AFT|VAL] {test specific arguments}"
+#define ENCAPS_USAGE "acvp_kyber{lvl} encapDecap AFT encaps ek=HEX m=HEX"
+#define DECAPS_USAGE "acvp_kyber{lvl} encapDecap VAL decaps dk=HEX c=HEX"
+#define KEYGEN_USAGE "acvp_kyber{lvl} keyGen AFT z=HEX d=HEX"
+
+typedef enum { encapDecap, keyGen } acvp_mode;
+
+typedef enum { AFT, VAL } acvp_type;
+
+typedef enum { encapsulation, decapsulation } acvp_encapDecap_function;
+
+/* Decode hex character [0-9A-Fa-f] into 0-15 */
+static unsigned char decode_hex_char(char hex) {
+  if (hex >= '0' && hex <= '9') {
+    return (unsigned char)(hex - '0');
+  } else if (hex >= 'A' && hex <= 'F') {
+    return 10 + (unsigned char)(hex - 'A');
+  } else if (hex >= 'a' && hex <= 'f') {
+    return 10 + (unsigned char)(hex - 'a');
+  } else {
+    return 0xFF;
+  }
+}
+
+static int decode_hex(const char *prefix, unsigned char *out, size_t out_len,
+                      const char *hex) {
+  size_t hex_len = strlen(hex);
+  size_t prefix_len = strlen(prefix);
+
+  // Check that hex starts with `prefix=`
+  // Use memcmp, not strcmp
+  if (hex_len < prefix_len + 1 || memcmp(prefix, hex, prefix_len) != 0 ||
+      hex[prefix_len] != '=') {
+    goto hex_usage;
+  }
+
+  hex += prefix_len + 1;
+  hex_len -= prefix_len + 1;
+
+  if (hex_len != 2 * out_len) {
+    goto hex_usage;
+  }
+
+  for (size_t i = 0; i < out_len; i++, hex += 2, out++) {
+    unsigned hex0 = decode_hex_char(hex[0]);
+    unsigned hex1 = decode_hex_char(hex[1]);
+    if (hex0 == 0xFF || hex1 == 0xFF) {
+      goto hex_usage;
+    }
+
+    *out = (hex0 << 4) | hex1;
+  }
+
+  return 0;
+
+hex_usage:
+  fprintf(stderr,
+          "Argument %s invalid: Expected argument of the form '%s=HEX' with "
+          "HEX being a hex encoding of %u bytes\n",
+          hex, prefix, (unsigned)out_len);
```

Comment:
maybe you onlt want to print the two chars that can't be parsed rather than the full hex string?

Reply:
Hm, maybe taste. I think it's easier to identify for the user what the error message is about if you give the full argument.